api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2383,6 +2386,240 @@ public void copyObject(String bucketName, String objectName, Map<String,String>
+      }
+
+      long size = stat.length();
+      if (src.length() != null) {
+        size = src.length();
+      } else {
+        if (src.offset() != null) {
+          size -= src.offset();
+        }
+      }
+
+      if (size < MIN_MULTIPART_SIZE && sources.size() != 1 && i != (sources.size() - 1)) {
+        throw new InvalidArgumentException("source " + src.bucketName() + "/" + src.objectName() + ": size "
+          + size + " must be greater than " + MIN_MULTIPART_SIZE);
+      }
+
+      objectSize += size;
+      if (objectSize > MAX_OBJECT_SIZE) {
+        throw new InvalidArgumentException("Destination object size must be less than " + MAX_OBJECT_SIZE);
+      }
+
+      if (size > MAX_PART_SIZE) {
+        long count = size / MAX_PART_SIZE;
+        long lastPartSize = size - (count * MAX_PART_SIZE);
+        if (lastPartSize > 0) {
+          count++;
+        } else {
+          lastPartSize = MAX_PART_SIZE;
+        }
+
+        if (lastPartSize < MIN_MULTIPART_SIZE && sources.size() != 1 && i != (sources.size() - 1)) {
+          throw new InvalidArgumentException("source " + src.bucketName() + "/" + src.objectName() + ": "
+            + "for multipart split upload of " + size
+            + ", last part size is less than " + MIN_MULTIPART_SIZE);
+        }
+
+        partsCount += (int)count;
+      } else {
+        partsCount++;
+      }
+
+      if (partsCount > MAX_MULTIPART_COUNT) {
+        throw new InvalidArgumentException("Compose sources create more than allowed multipart count "
+          + MAX_MULTIPART_COUNT);
+      }
+    }
+
+    if (partsCount == 1) {
+      ComposeSource src = sources.get(0);
+      if (headerMap == null) {
+        headerMap = new HashMap<>();
+      }
+      if ((src.offset() != null) && ( src.length() == null)) {
+        headerMap.put("x-amz-copy-source-range", "bytes=" + src.offset() + "-" );
+      }
+
+      if ((src.offset() != null) && ( src.length() != null)) {
+        headerMap.put("x-amz-copy-source-range", "bytes=" + src.offset() + "-" + (src.offset() + src.length() - 1));
+      }
+      copyObject(bucketName, objectName, headerMap, sse, src.bucketName(), src.objectName(), src.sse(),
+          src.copyConditions());
+      return;
+    }
+
+    Map<String, String> sseHeaders = null;
+    if (sse != null) {
+      sseHeaders = sse.headers();
+      if (headerMap == null) {
+        headerMap = new HashMap<>();
+      }
+      headerMap.putAll(sseHeaders);
+    }
+
+    String uploadId = initMultipartUpload(bucketName, objectName, headerMap);
+
+    int partNumber = 0;
+    Part[] totalParts = new Part[partsCount];
+    try {
+      for (int i = 0; i < sources.size(); i++) {
+        ComposeSource src = sources.get(i);
+
+        long size = src.objectSize();
+        if (src.length() != null) {
+          size = src.length();
+        } else {
+          if (src.offset() != null) {
```

Comment:
Just a minor comment:
You can combine these 2 lines:
```
} else {
   if (src.offset() != null) {
```
=> `} else if (src.offset() != null) {`